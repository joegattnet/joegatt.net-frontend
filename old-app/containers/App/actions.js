/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_NOTES,
  LOAD_NOTES_SUCCESS,
  LOAD_NOTES_ERROR,
} from './constants';

/**
 * Load the Notes, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_NOTES
 */
export function loadNotes() {
  return {
    type: LOAD_NOTES,
  };
}

/**
 * Dispatched when the Notes are loaded by the request saga
 *
 * @param  {array} notes The Note data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_NOTES_SUCCESS passing the notes
 */
export function NotesLoaded(notes, username) {
  return {
    type: LOAD_NOTES_SUCCESS,
    notes,
    username,
  };
}

/**
 * Dispatched when loading the Notes fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_NOTES_ERROR passing the error
 */
export function NoteLoadingError(error) {
  return {
    type: LOAD_NOTES_ERROR,
    error,
  };
}
