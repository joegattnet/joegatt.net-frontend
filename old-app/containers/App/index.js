/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React from 'react';
import Helmet from 'react-helmet';

import Header from 'components/Header';
import Page from 'components/Page';
import Content from 'components/Content';
import Footer from 'components/Footer';
import withProgressBar from 'components/ProgressBar';

export function App(props) {
  return (
    <Page>
      <Helmet
        titleTemplate="%s - React.js Boilerplate"
        defaultTitle="React.js Boilerplate"
        meta={[
          { name: 'description', content: 'A React.js Boilerplate application' },
        ]}
      />
      <Header />
      <Content>
        {React.Children.toArray(props.children)}
      </Content>
      <Footer />
    </Page>
  );
}

App.propTypes = {
  children: React.PropTypes.node,
};

export default withProgressBar(App);
