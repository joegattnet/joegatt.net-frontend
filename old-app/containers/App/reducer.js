/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_NOTES_SUCCESS,
  LOAD_NOTES,
  LOAD_NOTES_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentUser: false,
  userData: {
    Notes: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_NOTES:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['userData', 'Notes'], false);
    case LOAD_NOTES_SUCCESS:
      return state
        .setIn(['userData', 'Notes'], action.notes)
        .set('loading', false)
        .set('currentUser', action.username);
    case LOAD_NOTES_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
