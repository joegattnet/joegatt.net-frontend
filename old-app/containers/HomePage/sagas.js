/**
 * Gets the Notes of the user from Github
 */

import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_NOTES } from 'containers/App/constants';
import { NotesLoaded, NoteLoadingError } from 'containers/App/actions';

import request from 'utils/request';

/**
 * Github notes request/response handler
 */
export function* getNotes() {
  // Select username from store
  const requestURL = 'http://joegatt.net/texts.json';

  try {
    // Call our request helper (see 'utils/request')
    const notes = yield call(request, requestURL);
    yield put(NotesLoaded(notes));
  } catch (err) {
    yield put(NoteLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* githubData() {
  // Watches for LOAD_NOTES actions and calls getNotes when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_NOTES, getNotes);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  githubData,
];
