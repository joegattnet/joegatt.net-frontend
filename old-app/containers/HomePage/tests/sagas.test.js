/**
 * Tests for HomePage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_NOTES } from 'containers/App/constants';
import { NotesLoaded, NoteLoadingError } from 'containers/App/actions';

import { getNotes, githubData } from '../sagas';

const username = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('getNotes Saga', () => {
  let getNotesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getNotesGenerator = getNotes();

    const selectDescriptor = getNotesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getNotesGenerator.next(username).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the NotesLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First note',
    }, {
      name: 'Second note',
    }];
    const putDescriptor = getNotesGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(NotesLoaded(response, username)));
  });

  it('should call the NoteLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getNotesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(NoteLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = githubData();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_NOTES action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_NOTES, getNotes));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = githubDataSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = githubDataSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
