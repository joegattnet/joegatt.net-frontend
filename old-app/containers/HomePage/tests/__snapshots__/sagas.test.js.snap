exports[`getNotes Saga should call the NoteLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getNotes Saga should call the NoteLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/notes?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getNotes Saga should dispatch the NotesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getNotes Saga should dispatch the NotesLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/notes?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
